File: 08_leet_code/daily/n_queens.py
def n_queens(n, board=[]):
    if len(board) == n:
        return 1

    count = 0
    for col in range(n):
        board.append(col)
        if is_valid(board):
            count += n_queens(n, board)
        board.pop()
    return count

def is_valid(board):
    current_queen_row = len(board) - 1
    current_queen_col = board[-1]
    # Check if any queens can attack the last queen.
    for row, col in enumerate(board[:-1]):
        diff = abs(current_queen_col - col)
        if diff == 0 or diff == current_queen_row - row:
            return False
    return True

print(n_queens(4))

File: 08_leet_code/dynamic_programming/n_queens/n_queens.js
const nQueens = function(n) {
  const board = [];
  return rec(n, board);
};

const rec = function(n, board) {
  if (board.length === n) return 1;
  let count = 0;
  for (let i = 0; i < n; i++) {
    board.push(i);
    if (isValid(board)) {
      count += rec(n, board);
    }
    board.pop();
  }
  return count;
};

const isValid = function(board) {
  const currRow = board.length - 1;
  const currCol = board[board.length - 1];
  for (let r = 0; r < board.length - 1; r++) {
    const c = board[r];
    const diff = Math.abs(currCol - c);
    if (diff === 0 || diff === Math.abs(currRow - r)) {
      return false;
    }
  }
  return true;
};

console.log(nQueens(4));